key.ui/src/main/java/de/uka/ilkd/key/gui/SelectionHistory.java

```diff
@@ -0,0 +1,183 @@
+package de.uka.ilkd.key.gui;
+
+import de.uka.ilkd.key.core.KeYMediator;
+import de.uka.ilkd.key.core.KeYSelectionEvent;
+import de.uka.ilkd.key.core.KeYSelectionListener;
+import de.uka.ilkd.key.proof.Node;
+import de.uka.ilkd.key.proof.Proof;
+import de.uka.ilkd.key.proof.event.ProofDisposedEvent;
+import de.uka.ilkd.key.proof.event.ProofDisposedListener;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Traces the proof nodes selected by the user. Allows navigating forwards and backwards to
+ * show previous proof nodes.
+ *
+ * @author Arne Keller
+ */
+public class SelectionHistory implements KeYSelectionListener, ProofDisposedListener {
+    /**
+     * The KeY mediator.
+     */
+    private final KeYMediator mediator;
+    /**
+     * Previously selected nodes by the user.
+     */
+    private final Deque<Node> selectedNodes = new ArrayDeque<>();
```

Comment:
This is dangerous! You created a memory leak. 

Keeping proof nodes alive, results into the case where proofs are kept alive (reachable via a pointer in node.), and when proofs are reachable, everything belonging to a proof is reachable. Reachable objects are not removed by GC. 

You should either use `WeakReference`s or try to store these information inside the user data of a proof.

Reply:
Oh, thanks for the heads-up. I fixed it using `WeakReference`s.